### ISSUE TITLE
Dependency not found in ui test

### ISSUE THREAD

**Issue by rohitjoshi** [trailofbits/dylint#17]
I have implemented custom link to check if rsa generate() uses bits less than 2048. 
I did include `openssl` in the cargo dependencies including dev-dependencies but it doesn't find `openssl` crate when I run `cargo test`
```
use openssl::rsa::{Rsa, Padding};
fn main() {
    let rsa = Rsa::generate(1024).unwrap();
    let padding = Padding::PKCS1;
    let src: Vec<u8> = msg.into_bytes();
    let mut dest : Vec<u8> = vec![0;512];
    let _ = rsa.public_encrypt(&src, dest.as_mut_slice(), padding).unwrap();
}
```

**Comment by smoelius** [comment]
Thanks a lot for this report. Could you share your code so that I could try to reproduce the failure?

**Comment by smoelius** [comment]
Never mind. I am able to reproduce the bug. I am looking into it.

**Comment by rohitjoshi** [comment]
Thank you for dylint. I have been trying to figure out how to build custom links and dynamically load without modifying clippy.  Goal is to implement [OWASP top 10](https://cwe.mitre.org/data/definitions/1029.html). Let me know if you have done any work in this area.

**Comment by smoelius** [comment]
Thanks again for filing this issue. Here is a short term fix until I can find a good long term solution.

If you add `ui/main.rs` as `bin` target to your `Cargo.toml` file, you can run `cargo build -v` to see the flags passed to `rustc`. You can then extract the flags used for linking and pass those in `DYLINT_RUSTFLAGS`.

Specifically, I added this to my `Cargo.toml` file:
```toml
[[bin]]
name = "ui"
path = "ui/main.rs"
```
Running `cargo build -v`, I then saw this:
```sh
     Running `rustc --crate-name ui --edition=2018 ui/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C embed-bitcode=no -C split-debuginfo=unpacked -C debuginfo=2 -C metadata=4d4070dbb894c4b2 --out-dir /Users/sam.moelius/github/trailofbits/dylint-template/target/debug/deps -C linker=dylint-link -C incremental=/Users/sam.moelius/github/trailofbits/dylint-template/target/debug/incremental -L dependency=/Users/sam.moelius/github/trailofbits/dylint-template/target/debug/deps --extern clippy_utils=/Users/sam.moelius/github/trailofbits/dylint-template/target/debug/deps/libclippy_utils-4d989874afb6f43d.rlib --extern dylint_linting=/Users/sam.moelius/github/trailofbits/dylint-template/target/debug/deps/libdylint_linting-f23ae2b3c8af1f5b.rlib --extern if_chain=/Users/sam.moelius/github/trailofbits/dylint-template/target/debug/deps/libif_chain-295c1290813174bb.rlib --extern openssl=/Users/sam.moelius/github/trailofbits/dylint-template/target/debug/deps/libopenssl-16fe22e03505d944.rlib -L 'native=/usr/local/opt/openssl@1.1/lib'`
```
From that I extracted, the relevant `--edition`, `-L` flags, `--extern` flags. I then ran:
```sh
DYLINT_RUSTFLAGS="--edition=2018 -L dependency=/Users/sam.moelius/github/trailofbits/dylint-template/target/debug/deps --extern openssl=/Users/sam.moelius/github/trailofbits/dylint-template/target/debug/deps/libopenssl-16fe22e03505d944.rlib -L native=/usr/local/opt/openssl@1.1/lib" cargo test
```
And the test built and ran.

To reiterate: this is just a workaround. I intend to find a better long term solution. I'll leave the issue open until then.

> Goal is to implement OWASP top 10. Let me know if you have done any work in this area.

I haven't, but I think that is a fantastic idea.

**Comment by smoelius** [comment]
@rohitjoshi I think I've found a reasonable solution. Would you be able to verify that it solves your use case before I merge?

The solution is here: https://github.com/trailofbits/dylint/pull/20

In a nutshell, it adds two functions to `dylint_testing`: `ui_test_example` and `uit_test_examples`. I'll describe the former first.

One declares an example as one normally would in their `Cargo.toml` file, e.g.:
```toml
[[example]]
name = "ui"
path = "ui/main.rs"
```

Then, in their `lib.rs` file, they would use `ui_test_example` instead of `ui_test`, e.g., something like:
```rust
#[test]
fn ui() {
    dylint_testing::ui_test_example(env!("CARGO_PKG_NAME"), "ui");
}
```
where `"ui"` is the example's name.

`uit_test_examples` is similar, except it tests all declared examples, so there is no second argument.

There's a complete example here: https://github.com/trailofbits/dylint/tree/e767217b26376d7cf708e1bd9469776c68d79a2b/examples/question_mark_in_expression

Would you be able to try this out and see if it works for you?

Adding the following to your `Cargo.toml` file should allow you to use this version:
```toml
[patch.crates-io]
dylint_internal = { git = "https://github.com/smoelius/dylint" }
dylint_testing = { git = "https://github.com/smoelius/dylint" }
```

**Comment by smoelius** [comment]
Hey, @rohitjoshi. I've used #20 in some of my own work, and I'm reasonably confident it works how I want it to.

Still, do you want to give it a shot before I merge?

**Comment by smoelius** [comment]
I went ahead and merged #20 and pushed a new version. I think those changes address this issue, so I am going to close it.

Thanks again for this report, @rohitjoshi. Feel free to reopen the issue if you still have problems.

Also, please keep us updated on your OWASP top 10 project. Very interested to hear how it goes!

**Comment by rohitjoshi** [comment]
Sorry, I am on vacation . Will validate once back.

**Comment by smoelius** [comment]
> Sorry, I am on vacation . Will validate once back.

No worries. Enjoy!
